### PR TITLE
Integrate Travis-CI automated webpage generation and deploying

### PR DESCRIPTION
**Pull request by Syntaf**
I've been playing around with Travis-CI's container infrastructure a lot and think this could be a cool addition. I know back in DEVELOP @Jwely you mentioned if there was a way that the docs would auto update themselves upon commit. Sadly there currently doesn't exist any tools that do such, but on the bright side I took the liberty to write a tool that would do such for you guys! Check out [my fork](https://github.com/Syntaf/dnppy) for a working integration of this (the link in the readme.md won't work, since it's http://syntaf.github.io/dnppy/ under my fork) and see the automated [build logs](https://travis-ci.org/Syntaf/dnppy) here.
- [tl;dr](#tl;dr)
- [explanation](#explanation)
- [getting this working on your branch](#getting this working the master repo)

---

**note**: if you merge this pull request, make sure you commit a change that deletes the following lines in `.travis.yml` after you merge

```
env:
  global:
    secure: Wf3ESkJOuwC4Jwl6gGnXPhrLmmM5cLbo3OVVI7mdrXkUPE9le3Frlo6Dh9X+mrCCi+rxxWJ+98733pxZNF7gsmDfG1PFfBeQBxLuz9fnHsDZbfhhSH/AQ7A/cmkfaPXIzY/Thj0VbBx7s6fF7bKxlZGicrSE3pWVoqzTDho+SS/uluzcaYAJ2ru5dvi8knGoWZb9Eiw6my/NhSIbwge50pfruQPEY7I4+lBny/a8nItIWVmSS5q8dS3f7pglYqOG952NzZSqb2SLiHTWoES18Tv3zhjIuHTyoht724jnvQSqM9jw8bW5v/Co8NTMMFrXjcktAEJv3tDGAZ05Mz/zZH8Gu/YASsTZM5E10fuuqRaCzZh8PbTztGjhWqP1u7NMibxFSxQw4L8fe+Y1brwuWpPuxplwe3z7lY8R5V/sfReeebvm4MjOcMzAGdVmkZW4zOF30NpwPp8sroFYpM7yGli7BjmqiZsK2W95r+W7UrbDXIoM05pfEDQE4B9H0wifOsXE3AgmYK/EavJuRaTKUznu2a05ObJPvWaPP0k3RoU+jv7TozMlr8r1yameyv46tLyaqRjlxNhjlXkAZERNpygGsGlVmF27p7nf/FFsuMWEeXDwBlpbZI59g5frAtD48/FpwHaxmbJydMcgQSVyxwl0LFO9OXprbESlwR49Teo=
```

this is an auto generated encrypted key used for **my fork**, this needs to be deleted once merged before you do anything since this will not work for you.
## tl;dr

Simply put, this pull request will integrate Travis-CI's (free) remote building container infrastructure with your documentation chain. Every commit you make will proc a remote build, which will build your documentation and deploy it to your gh-pages. **there are some extra steps needed to get this working**
## explanation

[travis-ci](http://docs.travis-ci.com/user/getting-started/) is an awesome tool that lets you remotely build commits made to repositories, it's a free service and can really save you a lot of time. Once you activate a repo in their site, the repo needs a `.travis.yml` to receive the necessary build instructions . Here's what I told travis for dnppy:
- before anything, I specified `language`; dnppy will be built with `python - "2.7"`
- there's some small things we need to do before installing the required packages so we use `before_install`, we need to create a virtual environment for installing these packages to. Note travis CI uses container infrastructure, so we don't have sudo or any enviroment information available to us thus why we need to do this.
  - we grab miniconda which will create an env that we can install the packages to
- now onto `install`, we grab all the packages we need for python 2.7
- once `install` is finished we run the `script`, which builds our documentation
- if the script fails, `after_success` will **not** be called
- `after_success` will call `doc_upload.py` which will use the Github API token we supply travis to move the build docs into gh-pages
## getting this working on the master repo

There are two ways of going about this: 1. Give me temporary admin access until I finish setting this up for your guys, or 2. Follow the tutorial and i'll help troubleshoot. (2) is preferable because then if errors are encountered you can understand them yourselves.

The tutorial below is only from my memory, so bare with me if it doesn't work, I can troubleshoot with you if you send me information on what doesn't work
- head over to [https://travis-ci.org/](https://travis-ci.org/) and log in with your github
- go to the `accounts` tab and activate the dnppy repo, this will now trigger a build if a `.travis.yml` is detected. 
- This parts done now, next is travis configuration 

---
- Install ruby from [here](http://rubyinstaller.org/) , make sure you at least have `2.0.0`
- Open your command prompt
- Ensure ruby scripts are in your `PATH` and run `gem install travis -v 1.8.0 --no-rdoc --no-ri` (This requires admin elevation)
- Authenticate travis with `travis login`
- Head over to the settings tab of your profile and click Personal access tokens , then generate a new token img
- This token will give `doc_upload.py` permission to write to gh-pages, so copy this to your clipboard
- `cd` into the location of the dnppy repository
- run `travis encrypt GH_TOKEN=$YOUR_TOKEN --add env.global` where `$YOUR_TOKEN` is the token in your clipboard
- this should automatically add a couple lines to the bottom of the `.travis.yml` file, notably the `secure` value is a very long value now.
- upon pushing, this should trigger a build!

### CODE REVIEW

**Comment by Jwely** [comment]
You've been elevated, you should now have the access you need. Thanks for your contributions!

**Comment by Syntaf** [comment]
Look's like it's actually working with no extra configuration needed on my end, awesome! [link](https://travis-ci.org/NASA-DEVELOP/dnppy). Writing this script inspired me to start writing a [tool](https://github.com/Syntaf/travis-sphinx) that will allow anyone to start auto deploying their doc website through travis, so expect a new pull request soon that integrates my tool into `.travis.yml` and no longer need's that extra script!
